### PR TITLE
Refactoring to add more variations for the optimizer

### PR DESCRIPTION
**Pull request by watakandai**
estimated_imu

### CODE REVIEW

File: robotic_skin/calibration/error_functions.py [hunk 1]
```diff
@@ -0,0 +1,317 @@
...
+            Tpattern = joint_angle_func(curr_w, max_w, t)
+            # print(Tpattern.parameters, curr_w, max_w, t, d)
+            T = T.dot(Tpattern)
+
+    T = T.dot(Tdof2su)
+
+    return T.position
+
+
+def max_acceleration_joint_angle(curr_w, max_w, t):
```
Comment (peasant98) [inline robotic_skin/calibration/error_functions.py:147]: explain the function parameters a bit more here.

File: robotic_skin/calibration/error_functions.py [hunk 2]
```diff
@@ -0,0 +1,317 @@
...
+    max acceleration along a joint angle of robot.
+    """
+    # th_pattern = np.sign(t) * max_w / (curr_w) * (1 - np.cos(curr_w*t))
+    # th_pattern = np.sign(t) * max_w / (2*np.pi*C.PATTERN_FREQ) * (1 - np.cos(2*np.pi*C.PATTERN_FREQ*t))
+    th_pattern = max_w / (2*np.pi*C.PATTERN_FREQ) * np.sin(2*np.pi*C.PATTERN_FREQ*t) * t
+    # print('-'*20, th_pattern, curr_w, '-'*20)
+    return TransMat(th_pattern)
+
+
+def constant_velocity_joint_angle(curr_w, max_w, t):
```
Comment (peasant98) [inline robotic_skin/calibration/error_functions.py:158]: docstring

File: robotic_skin/calibration/error_functions.py [hunk 3]
```diff
@@ -0,0 +1,317 @@
...
+                # if p == 0:
+                #     print('[Dynamic Max Accel, %ith Joint]'%(d), n2s(max_accel_train), n2s(max_accel_model), curr_w, max_w)
+                error = np.sum(np.abs(max_accel_train - max_accel_model))
+                e2 += error
+                n_data += 1
+
+        return e2/n_data
+
+
+class ErrorFunctions():
```
Comment (peasant98) [inline robotic_skin/calibration/error_functions.py:361]: what will this be for?

File: robotic_skin/calibration/optimizer.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+        #     (4)    (2)     (4)
+        # dof -> dof -> vdof -> su
+        Tdof2vdof = TransMat(params[4:6])
+        Tvdof2su = TransMat(params[6:])
+
+    return Tdof2vdof.dot(Tvdof2su)
+
+
+class Optimizer():
+    """
```
Comment (peasant98) [inline robotic_skin/calibration/optimizer.py:33]: docstring here?

File: robotic_skin/calibration/plot_data.py [hunk 5]
```diff
@@ -0,0 +1,211 @@
...
+            [0, 0,      0.0825,     np.pi/2],
+            [0, 0.384,  -0.0825,    -np.pi/2],
+            [0, 0,      0,          np.pi/2],
+            [0, 0,      0.088,      np.pi/2]
+        ])
+
+    return dhparams
+
+
+def estimate_acceleration_analytically(Tdofs, Tjoints, Tdofi2su, d, i, curr_w):
```
Comment (peasant98) [inline robotic_skin/calibration/plot_data.py:69]: this is defined in another file; can you import it instead?

**Review by peasant98 — CHANGES_REQUESTED** [review]
Main thing is, could you add more docstrings, and also add some steps in readme how to run?

File: robotic_skin/calibration/optimizer.py [hunk 6]
```diff
@@ -31,9 +31,22 @@ def convert_dhparams_to_Tdof2su(params):
...
     """
+    Optimizer class to evaluate the data.
     """
     def __init__(self, error_functions, stop_condition=None, su_dhparams=None):
         """
+        Initializes the optimize with the following arguments:
+
+        Arguments
+        ---------
+        `error_functions`:
```
Comment (watakandai) [inline robotic_skin/calibration/optimizer.py:42]: dictionary

File: robotic_skin/calibration/utils.py [hunk 7]
```diff
@@ -376,6 +376,11 @@ def quaternion_l2_distance(q1, q2):
 
 
 def quaternion_from_two_vectors(source, target):
```
Comment (watakandai) [inline robotic_skin/calibration/utils.py:378]: we might not need this. We may need to delete this function

File: robotic_skin/calibration/error_functions.py (hunk 3 above)
Comment (watakandai) [reply robotic_skin/calibration/error_functions.py:361]: A error function base class

File: robotic_skin/calibration/plot_data.py (hunk 5 above)
Comment (watakandai) [reply robotic_skin/calibration/plot_data.py:69]: ah true. 
When I made this file, I didn't make the function that general. That's why I didn't import it from other file, but now I think we should

**Comment by peasant98** [comment]
ARM OF GOD

**Pull request merged**